radix-engine-tests/tests/wasm_metering.rs

```diff
@@ -185,3 +195,82 @@ fn test_grow_memory_out_of_cost_unit() {
     // Assert
     receipt.expect_specific_failure(is_costing_error)
 }
+
+#[test]
+fn test_max_instance_memory_exceeded() {
+    // Arrange
+    let mut test_runner = TestRunner::builder().build();
+
+    // Grow memory (wasm pages) to exceed default max wasm memory per instance.
+    let grow_value: usize = DEFAULT_MAX_WASM_MEM_PER_INSTANCE / (1024 * 64);
+
+    // Act
+    let code = wat2wasm(&include_str!("wasm/memory.wat").replace("${n}", &grow_value.to_string()));
+    let package_address = test_runner.publish_package(
+        code,
+        generate_single_function_abi(
+            "Test",
+            "f",
+            Type::Tuple {
+                element_types: vec![],
+            },
+        ),
+        BTreeMap::new(),
+        BTreeMap::new(),
+        AccessRules::new(),
+    );
+    let manifest = ManifestBuilder::new()
+        .lock_fee(FAUCET_COMPONENT, 10.into())
+        .call_function(package_address, "Test", "f", args!())
+        .build();
+    let receipt = test_runner.execute_manifest(manifest, vec![]);
+
+    // Assert
+    receipt.expect_specific_failure(|e| {
+        matches!(
+            e,
+            RuntimeError::ModuleError(ModuleError::TransactionLimitsError(
+                TransactionLimitsError::MaxWasmInstanceMemoryExceeded(_)
+            ))
+        )
+    })
+}
+
+#[test]
+fn test_max_transaction_memory_exceeded() {
+    // Test recursive instantiation of WASM with memory allocation of 1 MiB
+    // for each call (+additional memory for transaction execution).
+
+    // Arrange
+    let mut test_runner = TestRunner::builder().build();
+    let package_address = test_runner.compile_and_publish("tests/blueprints/recursion");
+
+    // Calculate value of additional bytes to allocate per call to exceed
+    // max wasm memory per transaction limit.
+    let grow_value: usize = DEFAULT_MAX_WASM_MEM_PER_TRANSACTION / DEFAULT_MAX_CALL_DEPTH;
+
+    // Ensure calculated wasm instance grow value is lower than instance limit.
+    assert!(grow_value <= DEFAULT_MAX_WASM_MEM_PER_INSTANCE);
+
+    // Act
+    let manifest = ManifestBuilder::new()
+        .lock_fee(FAUCET_COMPONENT, 10.into())
+        .call_function(
+            package_address,
+            "Caller",
+            "recursive_with_memory",
+            args!(DEFAULT_MAX_CALL_DEPTH as u32, grow_value),
+        )
+        .build();
+    let receipt = test_runner.execute_manifest(manifest, vec![]);
+
+    // Assert
+    receipt.expect_specific_failure(|e| {
+        matches!(
+            e,
+            RuntimeError::ModuleError(ModuleError::TransactionLimitsError(
+                TransactionLimitsError::MaxWasmMemoryExceeded(_)
+            ))
+        )
+    })
```

Comment:
Can we also add a test to assert the exact number of wasm memory usage? I want to make our memory measurements are implementation-independent and works deterministically between wasmer and wasmi.

Reply:
Added exact values of exceeded memory calculation and validation. Same values for wasmi and wasmer.